File: kotlin_client/src/main/kotlin/org/ezauth/client/Buckets.kt
package org.ezauth.client

import kotlinx.serialization.json.JsonPrimitive

/** Bucket and object storage operations. */
class Buckets internal constructor(private val client: BaseClient) {

    /** Object management sub-namespace. */
    val objects = Objects(client)

    suspend fun create(name: String): BucketResponse {
        val body = jsonObject("name" to JsonPrimitive(name))
        val resp = client.fetch("/v1/buckets", method = "POST", body = body, auth = AuthMode.AUTO)
        return client.json.decodeFromJsonElement(BucketResponse.serializer(), resp!!)
    }

    suspend fun list(): BucketListResponse {
        val resp = client.fetch("/v1/buckets", auth = AuthMode.AUTO)
        return client.json.decodeFromJsonElement(BucketListResponse.serializer(), resp!!)
    }

    suspend fun get(bucketId: String): BucketResponse {
        val resp = client.fetch("/v1/buckets/${bucketId.urlEncode()}", auth = AuthMode.AUTO)
        return client.json.decodeFromJsonElement(BucketResponse.serializer(), resp!!)
    }

    suspend fun delete(bucketId: String) {
        client.fetch("/v1/buckets/${bucketId.urlEncode()}", method = "DELETE", auth = AuthMode.AUTO)
    }
}

/** Object CRUD operations. */
class Objects internal constructor(private val client: BaseClient) {

    suspend fun put(
        bucketId: String,
        key: String,
        data: ByteArray,
        contentType: String = "application/octet-stream",
        userId: String? = null,
    ): ObjectResponse {
        val query = mutableMapOf<String, String?>()
        if (userId != null) query["user_id"] = userId
        val resp = client.fetchRaw(
            "/v1/buckets/${bucketId.urlEncode()}/objects/$key",
            method = "PUT",
            rawBody = data,
            contentType = contentType,
            query = query.ifEmpty { null },
        )
        resp.use { r ->
            val body = r.body?.string() ?: throw EZAuthError("Empty response", 0)
            val json = client.json.parseToJsonElement(body)
            return client.json.decodeFromJsonElement(ObjectResponse.serializer(), json)
        }
    }

    /** Downloads an object. Returns a pair of (data, contentType). */
    suspend fun get(
        bucketId: String,
        key: String,
        userId: String? = null,
    ): Pair<ByteArray, String> {
        val query = mutableMapOf<String, String?>()
        if (userId != null) query["user_id"] = userId
        val resp = client.fetchRaw(
            "/v1/buckets/${bucketId.urlEncode()}/objects/$key",
            method = "GET",
            contentType = "",
            query = query.ifEmpty { null },
        )
        resp.use { r ->
            val ct = r.header("Content-Type") ?: "application/octet-stream"
            val bytes = r.body?.bytes() ?: ByteArray(0)
            return bytes to ct
        }
    }

    suspend fun delete(
        bucketId: String,
        key: String,
        userId: String? = null,
    ) {
        val query = mutableMapOf<String, String?>()
        if (userId != null) query["user_id"] = userId
        client.fetchRaw(
            "/v1/buckets/${bucketId.urlEncode()}/objects/$key",
            method = "DELETE",
            contentType = "",
            query = query.ifEmpty { null },
        ).close()
    }

    suspend fun list(
        bucketId: String,
        userId: String? = null,
        limit: Int? = null,
        cursor: String? = null,
    ): ObjectListResponse {
        val query = mutableMapOf<String, String?>()
        if (userId != null) query["user_id"] = userId
        if (limit != null) query["limit"] = limit.toString()
        if (cursor != null) query["cursor"] = cursor
        val resp = client.fetch(
            "/v1/buckets/${bucketId.urlEncode()}/objects",
            auth = AuthMode.AUTO,
            query = query.ifEmpty { null },
        )
        return client.json.decodeFromJsonElement(ObjectListResponse.serializer(), resp!!)
    }
}

File: javascript_client/src/buckets.js
class Objects {
  constructor(client) {
    this._client = client;
  }

  async put(bucketId, key, data, { contentType = 'application/octet-stream', userId } = {}) {
    const query = {};
    if (userId !== undefined) query.user_id = userId;
    const resp = await this._client._fetchRaw(
      `/v1/buckets/${encodeURIComponent(bucketId)}/objects/${key}`,
      {
        method: 'PUT',
        body: data,
        contentType,
        query: Object.keys(query).length ? query : undefined,
      },
    );
    return resp.json();
  }

  async get(bucketId, key, { userId } = {}) {
    const query = {};
    if (userId !== undefined) query.user_id = userId;
    const resp = await this._client._fetchRaw(
      `/v1/buckets/${encodeURIComponent(bucketId)}/objects/${key}`,
      { query: Object.keys(query).length ? query : undefined },
    );
    const contentType = resp.headers.get('content-type') || 'application/octet-stream';
    const arrayBuffer = await resp.arrayBuffer();
    return { data: arrayBuffer, contentType };
  }

  async delete(bucketId, key, { userId } = {}) {
    const query = {};
    if (userId !== undefined) query.user_id = userId;
    await this._client._fetchRaw(
      `/v1/buckets/${encodeURIComponent(bucketId)}/objects/${key}`,
      {
        method: 'DELETE',
        query: Object.keys(query).length ? query : undefined,
      },
    );
  }

  list(bucketId, { userId, limit, cursor } = {}) {
    const query = {};
    if (userId !== undefined) query.user_id = userId;
    if (limit !== undefined) query.limit = limit;
    if (cursor !== undefined) query.cursor = cursor;
    return this._client._fetch(
      `/v1/buckets/${encodeURIComponent(bucketId)}/objects`,
      { auth: 'auto', query: Object.keys(query).length ? query : undefined },
    );
  }
}

export class Buckets {
  constructor(client) {
    this._client = client;
    this.objects = new Objects(client);
  }

  create({ name }) {
    return this._client._fetch('/v1/buckets', {
      method: 'POST',
      body: { name },
      auth: 'auto',
    });
  }

  list() {
    return this._client._fetch('/v1/buckets', { auth: 'auto' });
  }

  get(bucketId) {
    return this._client._fetch(`/v1/buckets/${encodeURIComponent(bucketId)}`, { auth: 'auto' });
  }

  delete(bucketId) {
    return this._client._fetch(`/v1/buckets/${encodeURIComponent(bucketId)}`, {
      method: 'DELETE',
      auth: 'auto',
    });
  }
}
